### PR TITLE
Datasets

### PR DESCRIPTION
**Pull request by abigail-rose**
Addition of code that allows the use of the CIFAR-100 dataset. 

- Hyperparameter 'dataset' added, which can be set to CIFAR (or 'Omniglot' to use Omniglot dataset).
- _base.py_ has been adapted from torchmeta, allowing efficient loading of CIFAR-100 using h5py.
- _cifar_one_shot_dataset.py_ creates dataset using CIFAR-100 images

### CODE REVIEW

File: lake/oneshot_cls.py [hunk 1]
```diff
@@ -168,16 +187,31 @@ def main():
...
+                               transform=image_tfms, target_transform=None),
+        batch_size=config['study_batch_size'], shuffle=False)
+
+    recall_loader = torch.utils.data.DataLoader(
+        OmniglotOneShotDataset('./data', train=False, download=True,
+                               transform=image_tfms, target_transform=None),
+        batch_size=config['study_batch_size'], shuffle=False)
+  
+  elif dataset_name == 'CIFAR':
+    np.random.seed(28)
```
Comment (affogato) [inline lake/oneshot_cls.py]: We should do this once for the whole program.
That way, we can make the seed a hyperparameter, and reliably reproduce our results.

File: lake/oneshot_cls.py [hunk 2]
```diff
@@ -168,16 +187,31 @@ def main():
...
-  study_loader = torch.utils.data.DataLoader(
-    OmniglotOneShotDataset('./data', train=True, download=True,
-                           transform=image_tfms, target_transform=None),
-    batch_size=config['study_batch_size'], shuffle=False)
-
-  recall_loader = torch.utils.data.DataLoader(
-    OmniglotOneShotDataset('./data', train=False, download=True,
-                           transform=image_tfms, target_transform=None),
-    batch_size=config['study_batch_size'], shuffle=False)
+  if dataset_name == 'Omniglot':
```
Comment (affogato) [inline lake/oneshot_cls.py:193]: It looks like everything is the same for all of these 'if' conditions, except for the Dataset.
I wonder if it would be more concise to load the appropriate dataset based on the condition, then just have one place to instantiate the DataLoader.
Thoughts?

File: lake/one_shot_dataset.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    labels = []
+
+    for r in range(1, self.num_runs + 1):
+      rs = str(r)
+      if len(rs) == 1:
+        rs = '0' + rs
+
+      target_path = os.path.join(self.root, self.target_folder)
+      # run_path = os.path.join(target_path, run_folder)
+      '''
```
Comment (affogato) [inline lake/one_shot_dataset.py]: why are there sections of code commented out here?

File: lake/cifar_one_shot_dataset.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        a_class = self.classes[i]
+
+        # selection of image
+        if self.mode == 'train':
+          selection = np.random.randint(low=0, high=300)
+        else:
+          selection = np.random.randint(low=300, high=600)
+
+        images.append(dataset[a_class][selection][0])  # PIL image
+        #labels.append(dataset[a_class][selection][1])  # 'fine' class label
```
Comment (affogato) [inline lake/cifar_one_shot_dataset.py:124]: Should we use the label, and convert to a number later? So that we still have the label if we want it.

File: lake/one_shot_dataset.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      x = torchvision.transforms.ToPILImage()(x)
+      x = torchvision.transforms.functional.affine(x, 0, translation, scale=1.0, shear=0, resample=Image.BILINEAR)
+
+      # Convert back to tensor
+      x = torchvision.transforms.functional.to_tensor(x)
+
+    return x
+
+
+class OneShotDataset(Dataset):
```
Comment (affogato) [inline lake/one_shot_dataset.py]: Is this class used?

**Review by affogato — CHANGES_REQUESTED** [review]
I've left some comments. Either make changes or we can organise a time to discuss.
Also, if including files/classes form another repo, let's put a reference at the top of the file to acknowledge where we got it from.

File: lake/oneshot_cls.py (hunk 2 above)
Comment (abigail-rose) [reply lake/oneshot_cls.py:193]: It actually isn't more concise to do this because we still have to load two datasets per condition (one for study and one for recall), and because each data loader takes different arguments (CIFAR data loader needs classes for our tests and also has an extra mode for validation) we still need to instantiate 2 datasets per condition. So if we change it to instantiate the data loader in one place, it ends up being more lines of code. 

Having said this, it may be more clear what is happening if the data loader is instantiated outside of the 'if' conditions. So I think it is still worth making this change, not for conciseness but for clarity.

File: lake/one_shot_dataset.py (hunk 5 above)
Comment (abigail-rose) [reply lake/one_shot_dataset.py]: I will delete this class, it was an idea that I had that I can implement later (this code doesn't really do anything). 

The idea is that there is one dataset class for all datasets. Will be a bit complicated to implement but if we are thinking of introducing more datasets later on it will make the code for oneshot_cls much more concise.

File: lake/cifar_one_shot_dataset.py (hunk 4 above)
Comment (abigail-rose) [reply lake/cifar_one_shot_dataset.py:124]: PyTorch DataLoader only works with numerical labels (specifically, ordered integers 0..n with n being number of classes), so as long as we are using PyTorch there isn't really a use in having the labels here -- means we can't use the dataset directly without adding extra code every time. Could add another array to save the labels if you want to be able to access them later.
